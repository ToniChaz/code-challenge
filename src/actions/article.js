import type from '../constants/actions/article'

export default {
  fetchArticles: () => ({ type: type.FETCH_ARTICLES }),
  fetchArticlesSuccess: articles => ({ type: type.FETCH_ARTICLES_SUCCESS, articles }),
  fetchArticlesError: error => ({ type: type.FETCH_ARTICLES_ERROR, error })
}
