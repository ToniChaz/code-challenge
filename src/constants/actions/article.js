export default {
  FETCH_ARTICLES: 'FETCH_ARTICLES',
  FILTER_ARTICLE: 'FILTER_ARTICLE',
  DELETE_ARTICLE: 'DELETE_ARTICLE',
  UPDATE_ARTICLE: 'UPDATE_ARTICLE',
  CREATE_ARTICLE: 'CREATE_ARTICLE',
  DELETE_ARTICLE_SUCCESS: 'DELETE_ARTICLE_SUCCESS',
  FETCH_ARTICLES_SUCCESS: 'FETCH_ARTICLES_SUCCESS',
  DELETE_ARTICLES_SUCCESS: 'DELETE_ARTICLES_SUCCESS',
  UPDATE_ARTICLE_SUCCESS: 'UPDATE_ARTICLE_SUCCESS',
  CREATE_ARTICLE_SUCCESS: 'CREATE_ARTICLE_SUCCESS'
}
