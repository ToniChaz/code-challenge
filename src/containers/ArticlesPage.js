import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import Articles from '../components/Articles'
import articleActions from '../actions/article'

export function mapStateToProps(state) {
  return { ...state.articles, ...state.router }
}

export function mapDispatchToProps(dispatch) {
  return bindActionCreators(articleActions, dispatch)
}

export default connect(mapStateToProps, mapDispatchToProps)(Articles)
