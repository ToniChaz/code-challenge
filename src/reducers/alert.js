import type from '../constants/actions/alert'

export function initialState() {
  return {
    visible: false,
    alertData: {
      color: 'primary',
      text: ''
    }
  }
}

function openAlert(state, { color, text }) {
  return {
    ...state,
    visible: true,
    alertData: {
      color,
      text
    }
  }
}

function closeAlert(state) {
  return {
    ...state,
    visible: false,
    alertData: initialState().alertData
  }
}

export default function(state = initialState(), action) {
  switch (action.type) {
    case type.OPEN_ALERT:
      return openAlert(state, action)
    case type.CLOSE_ALERT:
      return closeAlert(state, action)
    default:
      return state
  }
}
