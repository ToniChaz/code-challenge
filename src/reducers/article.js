import type from '../constants/actions/article'

export function initialState() {
  return {
    articles: null,
    loading: true,
    error: null
  }
}

function fetchArticles(state) {
  return {
    ...state,
    loading: true,
    error: null
  }
}

function fetchArticlesSuccess(state, { articles }) {
  return {
    ...state,
    articles,
    loading: false,
    error: null
  }
}

function fetchArticlesError(state, { error }) {
  return {
    ...state,
    loading: false,
    error
  }
}

export default function(state = initialState(), action) {
  switch (action.type) {
    case type.FETCH_ARTICLES:
      return fetchArticles(state, action)
    case type.FETCH_ARTICLES_SUCCESS:
      return fetchArticlesSuccess(state, action)
    case type.FETCH_ARTICLES_ERROR:
      return fetchArticlesError(state, action)
    default:
      return state
  }
}
