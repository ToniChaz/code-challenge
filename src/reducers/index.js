import { combineReducers } from 'redux'
import { routerReducer } from 'react-router-redux'
import articles from './article'

const rootReducer = combineReducers({
  articles,
  router: routerReducer
})

export default rootReducer
