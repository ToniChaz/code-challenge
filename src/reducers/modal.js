import type from '../constants/actions/modal'

export function initialState() {
  return {
    visible: false,
    edit: false,
    modalData: {
      author: '',
      content: '',
      title: '',
      published: '',
      tags: ''
    }
  }
}

function openModal(state, { modalData }) {
  return {
    ...state,
    visible: true,
    edit: !!modalData,
    modalData: modalData || initialState().modalData
  }
}

function closeModal(state) {
  return {
    ...state,
    visible: false,
    modalData: initialState().modalData
  }
}

function onChangeHandler(state, { name, value }) {
  const modalDataValues = { ...state.modalData, ...{ [name]: value } }

  return {
    ...state,
    modalData: modalDataValues
  }
}

export default function(state = initialState(), action) {
  switch (action.type) {
    case type.OPEN_MODAL:
      return openModal(state, action)
    case type.CLOSE_MODAL:
      return closeModal(state, action)
    case type.ON_CHANGE_HANDLER:
      return onChangeHandler(state, action)
    default:
      return state
  }
}
