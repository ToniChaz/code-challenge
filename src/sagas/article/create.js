import { call, put, takeEvery } from 'redux-saga/effects'
import { push } from 'react-router-redux'
import articleService from '../../services/article/create'
import action from '../../actions/article'
import alertAction from '../../actions/alert'
import modalAction from '../../actions/modal'
import type from '../../constants/actions/article'
import { CREATE } from '../../constants/service/mutations'

export function* createArticle({ article }) {
  try {
    const response = yield call(articleService, CREATE, article)
    yield put(action.createArticleSuccess(response.data.createArticle))
    yield put(alertAction.openAlert('success', 'The article has been saved successfully'))
    yield put(modalAction.closeModal())
    yield put(push('/'))
  } catch (error) {
    yield put(alertAction.openAlert('danger', `An error occurred: ${error.message}`))
  }
}

export function* watchCreateArticle() {
  yield takeEvery(type.CREATE_ARTICLE, createArticle)
}
