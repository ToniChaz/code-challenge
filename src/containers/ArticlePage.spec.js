import { mapDispatchToProps } from './ArticlesPage'
import { initialState } from '../reducers/article'

describe('ArticlesPage container', () => {
  test('should map ArticlesPage state props', () => {
    const state = initialState()
    expect(state).toMatchSnapshot()
  })

  test('should map ArticlesPage actions creators in a prop called "actions"', () => {
    const myMock = jest.fn()
    const mappedActions = mapDispatchToProps(myMock)

    const actionsPropNames = Object.keys(mappedActions)

    expect(actionsPropNames).toMatchSnapshot()
  })
})
