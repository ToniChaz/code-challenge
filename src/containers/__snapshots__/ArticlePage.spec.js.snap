exports[`ArticlesPage container should map ArticlesPage actions creators in a prop called "actions" 1`] = `
Array [
  "fetchArticles",
  "filterArticle",
  "deleteArticle",
  "updateArticle",
  "createArticle",
  "fetchArticlesSuccess",
  "deleteArticleSuccess",
  "updateArticleSuccess",
  "createArticleSuccess",
]
`;

exports[`ArticlesPage container should map ArticlesPage state props 1`] = `
Object {
  "article": null,
  "articles": null,
  "error": null,
  "loading": true,
}
`;
